tools/internal/autoupdate/registry.go

```diff
@@ -0,0 +1,348 @@
+	default:
+		return nil, fmt.Errorf("unrecognized registry: %s", registry)
+	}
+	req, err := http.NewRequest(http.MethodGet, registryUrl, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create request: %w", err)
+	}
+	if token != "" {
+		req.Header.Set("Authorization", "Bearer "+token)
+	}
+	req.URL.RawQuery = params.Encode()
+	return req, nil
+}
+
+func (r *Registry) getAllFromImage() (registry string, namespace string, repository string) {
+	image := r.Images[0].SourceImage
+	splittedImage := strings.Split(image, "/")
+
+	// Case 1: Handle default Docker Hub images like "flannel/flannel"
+	if len(splittedImage) == 2 && !strings.Contains(splittedImage[0], ".") {
+		return "dockerhub", splittedImage[0], splittedImage[1]
+	}
+
+	// Case 2: Handle images with a registry but no namespace like "k8s.gcr.io/pause"
+	if len(splittedImage) == 2 && strings.Contains(splittedImage[0], ".") {
+		return splittedImage[0], "", splittedImage[1]
+	}
+
+	// Case 3: Handle images with long paths like "gcr.io/cloud-provider-vsphere/csi/release/syncer"
+	if len(splittedImage) > 3 && strings.Contains(splittedImage[0], ".") {
+		return splittedImage[0], splittedImage[1], strings.Join(splittedImage[2:], "/")
+	}
+
+	// Default Case: Handle standard 3-part images like "quay.io/skopeo/stable"
+	return splittedImage[0], splittedImage[1], splittedImage[2]
+}
+
+func getSuseAuthToken(namespace, repository string) (string, error) {
+	tokenURL := "https://registry.suse.com/auth?service=SUSE+Linux+Docker+Registry"
+	req, _ := http.NewRequest("GET", tokenURL, nil)
+	params := req.URL.Query()
+	params.Add("scope", fmt.Sprintf("repository:%s/%s:pull", namespace, repository))
+	req.URL.RawQuery = params.Encode()
+
+	resp, err := doRequestWithRetries(req)
+	if err != nil {
+		return "", fmt.Errorf("suse token request failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	body, _ := io.ReadAll(resp.Body)
+	var tokenData SuseTokenResponse
+	if err := json.Unmarshal(body, &tokenData); err != nil {
+		return "", fmt.Errorf("failed to unmarshal suse token: %w", err)
+	}
+	return tokenData.AccessToken, nil
+}
+
+func doRequestWithRetries(req *http.Request) (*http.Response, error) {
+	maxRetries := 10
+	backoffFactor := 1 * time.Second
+	retryableStatuses := map[int]bool{
+		http.StatusInternalServerError: true, // 500
+		http.StatusBadGateway:          true, // 502
+		http.StatusServiceUnavailable:  true, // 503
+		http.StatusGatewayTimeout:      true, // 504
+	}
+
+	var resp *http.Response
+	var err error
+
+	for i := 0; i < maxRetries; i++ {
+		resp, err = httpClient.Do(req)
+		if err != nil {
+			return nil, fmt.Errorf("request failed: %w", err)
+		}
+
+		if resp.StatusCode >= 200 && resp.StatusCode < 300 {
+			return resp, nil
+		}
+
+		if !retryableStatuses[resp.StatusCode] {
+			return nil, fmt.Errorf("request failed with status %s", resp.Status)
+		}
+
+		if resp.Body != nil {
+			io.Copy(io.Discard, resp.Body)
+			resp.Body.Close()
+		}
+
+		if i == maxRetries-1 {
```

Comment:
It seems strange to have this check here when we are using a `for` loop that checks `maxRetries` at the beginning of every loop. Can this be simplified?